### DISCUSSION TITLE
What algorithm is using for nodeAutoColorBy?

### DISCUSSION THREAD

**Discussion by dzhai5** [vasturiano/react-force-graph#525] [Q&A] [answered]
I am working on a project that hopes to have a label bar telling different categories of the nodes get what different color. I assume nodeAutoColorBy is using some hash color method, but how do I find the source code or the specific function for this?

**Comment by vasturiano** [comment, marked as answer]
@dzhai5 if auto-coloring is enabled, nodes that don't already have a color associated get assigned one according to the criteria set in `nodeAutoColorBy`. 
The colors come from [d3's schemePaired](https://d3js.org/d3-scale-chromatic/categorical#schemePaired).

Relevant source code:
https://github.com/vasturiano/three-forcegraph/blob/d091b6f7d09a0fd35032cad560b19dcbfe76bd3f/src/forcegraph-kapsule.js#L611-L614

https://github.com/vasturiano/three-forcegraph/blob/master/src/utils/color-utils.js
